docs/decisions/0004-error-handling.md

```diff
@@ -0,0 +1,59 @@
+
+- **Missing original exception details**. Certain SK exceptions do not preserve the original failure or exception details and do not expose them through their properties. This omission prevents SK client code from understanding the problem and handling it properly.
+
+- **No consolidated way for exception handling**. In some cases, SK has an exception type per component implementation, making it impossible for SK client code to handle them in a consolidated manner. Instead of having a single catch block, SK client code needs to include a catch block for each component implementation. Moreover, SK client code needs to be updated every time a new component implementation is added or removed.
+
+## Solution
+
+To address the identified issues, the following steps can be taken:
+
+- **Exception propagation**: To enable SK exceptions to propagate to SK client code, all SK components that currently do not do so should be modified to rethrow the exception instead of suppressing it. This allows SK client code to use try/catch blocks for handling and catching exceptions.
+
+- **Improper exception usage**: Modify SK code to throw .NET standard exceptions, such as ArgumentOutOfRangeException or ArgumentNullException, when class argument values are not provided or are invalid, instead of throwing custom SK exceptions. Analyze SK exception usage to identify other potential areas where standard .NET exceptions can be used instead.
+
+- **Exception hierarchy inconsistency**: Refactor SK custom exceptions that are not derived from SKException to derive from SKException. This ensures consistency in the SK exception hierarchy.
+
+- **Missing original exception details**: Identify all cases where the original exception or failure is not preserved either as an inner exception or as a property/message of the rethrown exception, and address them. For example, AI service connectors that use HttClient should preserve the original status code by either setting it as an inner exception or including it as part of the rethrown exception's property.
+
+- **No consolidated way for exception handling**:
+  - [Memory connectors]: Introduce a new exception called VectorDbStorageException, along with derived exceptions like UpsertRecordException and RemoveRecordException, if additional details need to be captured and propagated to SK client code. Refactor all SK memory component implementations to use these exceptions instead of implementation-specific memory storage exceptions.
+  - [AI service connectors]: Restrict the usage of AIException to AI service connectors only and replace its usage in other SK components with more relevant exceptions. For cases where AIException is thrown, use HttpOperationException describved below as an inner exception to allow SK client code to access the original status code and exception/error details, if necessary.
```

Comment:
Regarding multiple memory stores issue, same applies for AI service connectors. Since it's possible today to register multiple AI services, as a result, we expect only `AIException` exception, but it's not clear how to handle each AI service exception separately. Also, I would probably rename `AIException` to `AIServiceException`.

Reply:
As usual, **well-thought-through** scenarios would help here.